### PR TITLE
feat: guidance complexity measurement and monitoring framework (#211)

### LINKED ISSUE #211: Guidance complexity management — symmetric retirement for guardrail additions
## Customer Problem

The Review Kaizen system (#148) creates a pipeline for detecting recurring defect patterns and proposing guardrail additions to prevent them. This is working as intended — but the pipeline is asymmetric: **guardrails accumulate with no symmetric mechanism to consolidate, simplify, or retire them.**

Over time, this produces three customer-experience risks:

### 1. Context Saturation

Every line of guidance competes for the LLM's attention. There is empirical evidence that LLMs follow instructions less reliably as context grows. At some point, adding a guardrail rule to prevent one class of error *causes* another class of error by diluting attention on existing rules — fighting the very phenomenon we're trying to fix.

### 2. Special-Case Accretion

The guardrail proposal pipeline routes fixes through `systemic_fix_type` categories (`instruction | skill | agent-prompt | plan-template`), all of which add rules *within the existing structure*. Each rule is rational in isolation; collectively they become a checklist no one — human or AI — can reliably follow.

### 3. Adopter Onboarding Burden

For anyone cloning this as a template, the complexity of the system is the first thing they experience. Every guardrail rule added to prevent a past issue raises the barrier for new adopters who never encountered that issue.

## Current State (Measured)

| Layer | Lines | Notes |
|---|---|---|
| Agent definitions (14 agents) | 3,823 | Code-Conductor: 696, Process-Review: 598, Code-Critic: 454 |
| Instructions (5 files) | 526 | Loaded per-agent |
| Skills (14 skills) | 1,581 | On-demand |
| copilot-instructions | 114 | Every conversation |
| Scripts | 2,497 | Runtime only |
| Contract tests | 3,211 | Dev-time only |
| **Total guidance** | **6,044** | Agents + instructions + skills + copilot-instructions |

The heaviest agents (Code-Conductor, Process-Review, Code-Critic) are the ones that will absorb most guardrail additions from the kaizen pipeline.

## Design Intent

**The system should *learn and simplify*, not just accumulate.**

The Deming philosophy that inspired #148 doesn't just prescribe "add more checks." The Toyota Production System's kaizen includes *simplification* — reducing waste, not just adding gates. The goal is to make the system produce better output structurally, not to catch more defects through longer checklists.

The adaptive depth system (#150) reduces *prosecution effort* for categories with low sustain rates, but it doesn't reduce the *volume of guidance* in agent files. A category at `skip` depth still has all its guardrail rules loaded into context. The guidance bloat remains.

The missing counterpart is: **when guardrails have served their purpose (the category consistently produces low sustain rates), how do we consolidate or retire the specific rules, keeping the system lean?**

A key distinction: *rule-level prevention* ("add a rule telling the agent not to do X") catches defects but doesn't make them impossible. *Structural prevention* (change the architecture so X can't happen) makes the defect impossible and often *removes* rules rather than adding them.

## User Segments

- **Active users** — experience context window pressure and attention dilution as guidance grows
- **Template adopters** — experience onboarding overwhelm from accumulated rules they don't need
- **Maintainers** — experience difficulty understanding interactions between many specific rules

### Current Journey (Active User)

1. Kaizen pipeline detects recurring defect pattern across PRs
2. Process-Review §4.9 proposes a guardrail rule addition (instruction/skill/agent-prompt)
3. Rule is added to the relevant agent or instruction file
4. Agent file grows longer; all conversations load the new rule
5. Over time, agent context fills with accumulated specific-case rules
6. Agent follows some rules less reliably as context grows (attention dilution)
7. New defect types emerge *because of* diluted attention on existing rules

### Target Journey (Active User)

1. Kaizen pipeline detects recurring defect pattern
2. **Structural Prevention Gate evaluates: can this defect class be caught earlier in the pipeline?**
3. If structural fix exists (lint rule, contract test, scaffold, workflow ordering) → propose that instead. Rule becomes unnecessary.
4. If rule-level fix is genuinely needed → propose it, tagged as `rule-level`
5. **Rule Compression monitors accumulated rules per section.** When density exceeds threshold, consolidation triggers — N specific rules become general principles + examples.
6. **Complexity-Triggered Retirement** flags rules where calibration data shows sustained improvement — compound signal (low sustain rate + skip depth + no regression) triggers consolidation review
7. Procedural sections that are rarely used move from always-loaded agent files to on-demand skills
8. Agent context stays lean; attention stays focused on principles that generalize to novel cases

### Current Journey (Template Adopter)

1. Clone repository
2. Open an agent file (e.g., Code-Critic) — 750+ lines
3. See a mix of general principles and very specific accumulated rules (e.g., `[ordered]@{}.Clone()` trap)
4. Overwhelmed by specificity that doesn't apply to their project
5. Unclear which rules are essential principles vs. accumulated patches

### Target Journey (Template Adopter)

1. Clone repository
2. Open an agent file — lean, principle-based guidance with worked examples
3. Specific project rules live in skills loaded on-demand, or in CI checks that run automatically
4. Onboarding reads as "good engineering principles" not "list of past bugs"

## Walkthrough-Vetted Principles

These principles were tested against real codebase rules and pending improvement issues (#47, #156, #201, #202) to verify they work in practice.

### Principle 1: Structural Prevention Gate — Catch Issues Further Upstream

**Before proposing a rule, ask: can this defect class be caught earlier in the pipeline?**

Structural prevention moves the catch point to a stage where the defect physically cannot pass through — a lint rule, a CI check, a template that generates correct code, or a workflow ordering that makes the correct sequence automatic. It's not about fixing the individual defect; it's about eliminating the *pathway* by which that class of defect reaches the stage where the rule operates.

**Upstream catch hierarchy** (earlier = better):
- **Template/scaffold** → prevents defect at creation time (before writing code)
- **Type system / API design** → prevents wrong usage at compile/write time
- **Contract test** → catches at dev time (before build)
- **CI lint rule** → catches at build time (before review)
- **Workflow ordering** → prevents wrong sequence at orchestration time (before delegation)
- **Rule-level fix** → catches at review time (last resort)

**Walkthrough evidence**: Code-Critic §6 has 7 specific PowerShell rules. At least 3 (exit-code checks, `.Clone()` trap, collection-return semantics) could be replaced by PSScriptAnalyzer custom rules or Pester contract tests that catch the defects at lint/CI time — making the Code-Critic checklist items redundant. Issue #156 (Bug Report Protocol) proposed adding a rule about investigation ordering; a structural fix would bake design-context loading into the workflow hook automatically.

**Pipeline modification**: Process-Review §4.9 (Guardrail Proposals) gains a pre-proposal gate: evaluate the upstream catch hierarchy before emitting a rule proposal. If a structural fix exists → propose that instead.

### Principle 2: Rule Compression — Specific Rules → General Principles

**When rules accumulate in the same section, consolidate into general principles + worked examples.**

Multiple specific rules often encode a single general principle. Principles generalize to novel cases; specific checklist items only catch the exact case they describe. Compression reduces line count while *improving* coverage because the LLM can apply a principle to situations the specific rule never anticipated.

**Walkthrough evidence**: Code-Critic §6's 7 rules compress to 3 principles: "validate all boundary interactions" (covers exit codes, injection, sanitization, regex), "cross-reference enumerated values with authoritative sources," and "verify .NET method availability and pipeline semantics." ~40% line reduction while increasing generalizability.

**Trigger**: When a section accumulates N+ independent rules in the same category (N to be calibrated — walkthrough suggests ~5-7 specific rules is a reasonable threshold for compression).

### Principle 3: Load-Time vs. On-Demand Complexity (Agent → Skill Extraction)

**Procedural sections that run occasionally should be skills (on-demand), not agent sections (always-loaded).**

Agent files are loaded into every conversation. Skills are loaded only when invoked. Moving rarely-used procedural workflows from agents to skills reduces always-loaded context without losing capability.

**Walkthrough evidence**: Process-Review §4.8 (Upstream Gotcha Lifecycle, ~120 lines) was identified as a potential extraction candidate. However, §4.8 is a new addition that hasn't had a fair chance — it silently skips in the upstream repo because no `local-gotchas.instructions.md` file exists yet. **Decision: widen §4.8's trigger to also run during calibration-only mode (matching §4.9), add invocation monitoring, and defer the extraction/retirement decision until monitoring data supports it.**

**Criteria for extraction**: Procedural sections that (a) have explicit skip conditions ("skip in subagent mode X"), (b) are self-contained workflows, (c) don't need to influence the agent's general behavior, and (d) have monitoring data showing low invocation frequency are candidates.

### Principle 4: Compound-Signal Retirement — Probation, Not Expiry

**Rules become retirement candidates only when multiple independent signals confirm the defect class is resolved.**

Pure time-based retirement is too blunt — "skip depth for 90 days" doesn't mean the defect is gone, just that prosecution stopped looking. The compound signal is:

1. **Low sustain rate** — the category consistently produces few sustained findings
2. **Skip prosecution depth** — the adaptive depth system has stopped active prosecution
3. **No regression** — when prosecution was absent, the defect class didn't resurface

All three signals must agree before retirement is proposed. This is a "probation period" model, not a half-life.

**Current state**: prosecution_depth_state doesn't exist yet because the calibration system has only 4 entries (~20 effective count needed per category for depth transitions). This mechanism will activate naturally as more calibration data accumulates.

### Principle 5: Complexity Budget Per Agent

**Each agent has a soft complexity ceiling. Adding a rule past the ceiling triggers mandatory consolidation.**

This creates the symmetric pressure described in the issue — you can't just keep adding rules; at some point you must simplify to make room. The ceiling is measured in **directive density** (count of must/shall/always/never/required directives + checklist items per agent), not raw line count, because a 50-line section with 15 independent obligations is more burdensome than a 100-line section with 3 clear principles.

**Mechanism**: A lightweight script counts actionable directives per agent file. When a proposed guardrail addition would push an agent past its budget, the proposal is flagged as "requires consolidation first."

### Principle 6: Measure What Matters — Sustain Rate, Not Just Line Count

**Consolidation is only valid if quality is maintained. Sustain rate regression after consolidation = rollback.**

The calibration system (#150) already tracks per-category sustain rates. After any consolidation (rule compression, retirement, agent→skill extraction), the sustain rate in affected categories must be monitored. If it regresses beyond a threshold, the consolidation is reversed — the specific rules are restored.

This is the safety net that makes all other principles safe to apply. Without measurement, consolidation is guesswork.

---

### Design Decisions (Confirmed)

| ID | Decision | Chosen Option | Rationale |
|----|----------|---------------|-----------|
| D1 | How to detect complexity | **Regex script with calibration override** — `measure-guidance-complexity.ps1` counts directive density per agent (must/shall/always/never/mandatory + checklist items), excluding code blocks; soft ceilings in `.copilot-tracking/guidance-complexity.json`; override comments for false positives; section count + nesting depth as supplemental signals | Automated, Pester-testable, integrates with quick-validate |
| D2 | How to trigger compression | **Complexity ceiling trigger** — when a section exceeds its soft ceiling, guardrail proposals targeting that agent require compression first | Creates symmetric pressure: simplify before adding |
| D3 | What retirement means | **Consolidation + archival** — compress to principles, move specifics to `Documents/Archive/retired-rules/` with metadata (source agent, section, version range, restoration instructions) | Knowledge preserved; context reduced; rollback easy |
| D4 | How to distinguish structural vs. rule-level | **Hybrid** — script defaults `prevention_level: unknown`; Process-Review §4.9 evaluates against upstream catch hierarchy and reclassifies to `structural` or `rule-level` | Deterministic default where possible; flexible reasoning for novel patterns |
| D5 | How to measure quality after consolidation | **Consolidation event tracking + regression threshold** — record events in calibration data; sustain rate monitoring with +10pp regression threshold; human review (not auto-rollback) | Uses existing infrastructure; regression detection is meaningful |
| D6 | How to extract agent sections to skills | **Skill reference stubs** — 2-3 line stub in agent ("what and when"), full procedure in skill ("how"); 4 extraction criteria defined; **framework only — no immediate extractions**; §4.8 gets widened trigger + monitoring to collect evidence for future extraction/retirement decision | Framework-first; candidates emerge from monitoring data |

### Implementation Phases

**Phase 1 — Data-independent (ship first):**
- **D1**: `measure-guidance-complexity.ps1` + `guidance-complexity.json` config + quick-validate integration
- **D2**: Compression trigger wiring in Process-Review §4.9 (ceiling flag referenced from D1 output)
- **D6**: Extraction criteria documented; §4.8 trigger widened to calibration-inclusive mode; invocation monitoring added

**Phase 2 — Data-dependent (ship after calibration matures, ~20+ effective count per category):**
- **D3**: Compound-signal retirement (requires prosecution_depth_state to exist)
- **D4**: Structural Prevention Gate in §4.9 (requires stabilized Sub C pipeline)
- **D5**: Consolidation event tracking + regression monitoring (requires consolidation events to exist)

**Rationale for phasing**: Phase 1 mechanisms work purely from file analysis and structural rules — no calibration data needed. Phase 2 mechanisms depend on sufficient calibration history to make meaningful decisions. Shipping Phase 1 first allows the system to start capping complexity immediately while calibration data accumulates for Phase 2.

### System Touchpoints

| Artifact | Change | Phase |
|---|---|---|
| `.github/scripts/measure-guidance-complexity.ps1` | **New** — directive density counting script | 1 |
| `.copilot-tracking/guidance-complexity.json` | **New** — soft ceilings config per agent | 1 |
| `copilot-instructions.md` (Quick-validate) | **Modified** — add D1 script to pre-merge check list | 1 |
| Process-Review §4.8 | **Modified** — widen trigger to calibration-only mode + monitoring | 1 |
| Process-Review §4.9 | **Modified** — compression-required flag when ceiling exceeded | 1 |
| `.github/scripts/aggregate-review-scores.ps1` | **Modified** — `prevention_level:`, `retirement_candidates:`, `consolidation_events[]` fields | 2 |
| `Documents/Archive/retired-rules/` | **New convention** — archive for retired rules with restoration metadata | 2 |

**Net-new artifacts capped**: 1 script + 1 config file (Phase 1). Phase 2 additions modify existing files only (plus archive convention).

### Testing Scope

| Change Type | Unit (Pester) | Integration | E2E |
|---|---|---|---|
| D1 script (directive counting) | ✅ | ❌ | ❌ |
| D2 (compression trigger in §4.9) | ✅ Contract test | ❌ | ❌ |
| D6 (§4.8 trigger widening) | ✅ | ❌ | ❌ |
| Phase 2 script changes | ✅ | ✅ (calibration pipeline) | ❌ |

**Integration test scenarios**:
- **[D1 + D2]**: `measure-guidance-complexity.ps1` detects an agent exceeds its ceiling → calibration report includes compression warning → Process-Review §4.9 flags "compression required before adding to [agent section]"
- **[D5 + D3]**: consolidation event recorded → N PRs later → sustain rate increases past +10pp threshold → regression flagged for human review

### Edge Cases

- **Empty calibration data**: D1/D2 work without calibration data (pure file analysis). D3/D4/D5 emit "insufficient data" and skip gracefully.
- **All agents under ceiling**: Compression trigger doesn't fire. System is passive until needed.
- **Regression after consolidation**: D5 detects via sustain rate threshold; human reviews and can restore from archive. Not auto-rollback.
- **§4.8 always skips**: With widened trigger, if it still always skips (no gotcha file), the output log provides data to inform future extraction/retirement decision.
- **No retirement candidates**: Compound signal (D3) only fires when all three conditions are met. If sustain rates stay high, no retirements are proposed — the system stays protective.

---

## Adversarial Challenge Summary

Three independent Code-Critic prosecution passes produced 21 unique findings after deduplication. Key findings and their dispositions:

| Finding | Severity | Disposition |
|---|---|---|
| Regex undercounts cognitive complexity (checklist items vs. directive words) | High | **Incorporated** — D1 refined: count checklist items as primary signal, exclude code blocks, supplement with section count + nesting depth |
| Behavioral leakage test missing after extraction | High | **Incorporated** — D6 extraction criteria include post-extraction Pester suite pass + invocation verification + stub trigger-condition contract test |
| Script integration breadth (new script = maintenance burden) | High | **Addressed** — net-new artifacts capped to 1 script + 1 config; Phase 2 modifies existing files only |
| Category mismatch between calibration data types and `$knownCategories` | High | **Implementation note** — audit `$knownCategories` list during implementation |
| No rollback procedure defined for retired rules | High | **Incorporated** — D3 archive convention includes metadata (source agent, section, version range, restoration instructions) |
| §4.9 runs every calibration cycle — contradicts "low frequency" extraction criterion | High | **Corrected** — §4.9 removed as extraction candidate; D6 scoped as framework only |
| Phase delivery: data-dependent vs. data-independent phases | High | **Incorporated** — Phase 1 (D1, D2, D6) / Phase 2 (D3, D4, D5) split |
| Directive counting should exclude code blocks and examples | Medium | **Incorporated** — D1 regex excludes fenced code blocks |
| Compression ratio can't be predicted from density alone | Medium | **Accepted** — D2 flags for human-guided compression, not auto-compress |
| §4.8 needs fair chance before extraction decision | Medium | **Incorporated** — widen trigger + monitor, defer extraction decision |

All findings were non-blocking (design prosecution, not gatekeeping). Full merge-deduplication was performed across 3 passes with cross-pass dedup.

---

## Rejected Alternatives

| Alternative | Why Rejected |
|---|---|
| Agent-only structural evaluation (D4 Option A) | Non-deterministic; no Pester-testable gate. Hybrid gives both deterministic script default and flexible agent reasoning. |
| Manual compression trigger (D2 Option A) | No systematic pressure. Relies on someone noticing density. Doesn't create symmetric pressure against accumulation. |
| Full deletion for retirement (D3) | Institutional knowledge lost. If defect recurs, re-proposed rules may be worse than originals. Archival preserves knowledge with restoration path. |
| Periodic time-based compression review (D2 Option C) | May trigger unnecessarily; adds periodic obligation. Ceiling-triggered compression fires only when it matters. |
| Implicit skill invocation via orchestrator (D6 Option B) | Coupling moves upstream; breaks direct invocation. Stub keeps agent self-aware of its capabilities. |
| §4.8 immediate extraction (D6) | New feature, needs fair chance. Widened trigger + monitoring provides evidence for future disposition. §4.8 hasn't had adequate usage data to justify extraction or retirement. |
| Pure time-based rule expiry (D3 half-life model) | "Skip depth for 90 days" doesn't mean the defect is gone, just that prosecution stopped looking. Compound-signal model requires multiple independent confirmations. |
| Complex weighted scoring for density (D1) | Over-engineered. Simple directive + checklist counting with section depth supplementals provides sufficient signal. |

---

## Acceptance Criteria

- [ ] `measure-guidance-complexity.ps1` exists and counts directive density per agent, excluding code blocks, with Pester test coverage
- [ ] `.copilot-tracking/guidance-complexity.json` defines soft ceilings per agent file
- [ ] Quick-validate includes the complexity check in pre-merge validation
- [ ] Process-Review §4.9 references complexity ceiling and flags "compression required" when exceeded (D2)
- [ ] Process-Review §4.8 trigger widened to also run in calibration-only mode, with invocation monitoring
- [ ] Extraction criteria for agent→skill moves are documented (D6 framework)
- [ ] Phase 2 design is documented and ready for implementation after calibration matures (~20+ effective count per category)
- [ ] Consolidation + archival convention defined with restoration metadata schema (D3)
- [ ] Structural prevention gate evaluation added to §4.9 proposal workflow (D4, Phase 2)
- [ ] Consolidation event tracking in calibration data with +10pp regression threshold (D5, Phase 2)
- [ ] Sustain rate monitoring validates quality is maintained after any consolidation
- [ ] Net-new maintenance burden is bounded (1 script + 1 config in Phase 1; Phase 2 modifies existing files)

## Scenarios (CE Gate Drafts)

1. **Functional — Structural Prevention**: When the guardrail proposal pipeline detects a recurring pattern, it evaluates the upstream catch hierarchy. For a defect class that can be caught by a CI lint rule, it proposes the lint rule instead of an agent checklist item.
2. **Functional — Rule Compression**: After compression of a high-density agent section (e.g., Code-Critic §6), the agent file has fewer lines and fewer independent directives while Pester tests and quick-validate checks continue to pass.
3. **Intent — Principled Reading**: An adopter cloning the template for a new project encounters agent guidance that reads as principled and general ("validate all boundary interactions") rather than as a list of specific past-incident patches.
4. **Functional — Retirement Safety Net**: After consolidation, the calibration system monitors sustain rates in affected categories and flags regression within N PRs.

## Customer Surface

Internal workflow (agent guidance files, calibration pipeline). CE Gate surface: `ce_gate: false` — process-only improvement with no customer-facing runtime surface.

## CE Gate Readiness Assessment

1. **Surface type**: Process-only (agent guidance files, scripts, calibration pipeline)
2. **Tool availability**: N/A — no runtime surface to exercise
3. **Manual fallback**: N/A
4. **Draft scenarios**: Scenarios 1–4 above verify through Pester tests, quick-validate checks, and calibration data analysis rather than UI/API interaction

## Related Issues

- #148 — Review Kaizen (parent of the guardrail addition pipeline)
- #149 — Root cause tagging (provides `systemic_fix_type` routing)
- #150 — Prosecution adaptive depth (reduces prosecution effort but not guidance volume)
- #151 — Guardrail proposal pipeline (the addition side this issue complements)
- #180, #181 — Token budget optimization (runtime concern; this issue addresses design-time guidance volume)
- #47, #156, #201, #202 — Pending improvement issues used as walkthrough test cases

### PR DESCRIPTION
**Pull request by Grimblaz**
## Summary

Implements Phase 1 of the guidance complexity management system (issue #211). Introduces a measurement script that counts directive density per agent file, wires the check into Process-Review's review pipeline, and integrates it into the Quick-validate gate. All agents are currently below the initial ceiling of 128 directives.

## Changed Files

| File | Change |
|---|---|
| `.github/scripts/measure-guidance-complexity.ps1` | NEW — directive density counter; JSON output; depth-aware fence tracking |
| `.github/config/guidance-complexity.json` | NEW — tunable ceiling config (default: 128) |
| `.github/scripts/Tests/measure-guidance-complexity.Tests.ps1` | NEW — 33 unit tests (all GREEN, +6 fence edge case tests from GH review) |
| `.github/scripts/Tests/guidance-complexity-contract.Tests.ps1` | NEW — 7 sync-check contract tests |
| `Documents/Design/guidance-complexity.md` | NEW — design doc (D1-D6, Phase 2 criteria, Rule Compression Approach, Ceiling Management) |
| `.github/agents/Process-Review.agent.md` | MODIFIED — §4.7 Step 1b, §4.8 trigger, §4.9 wiring; try/catch + .error field |
| `.github/copilot-instructions.md` | MODIFIED — complexity check added to Quick-validate |
| `.github/architecture-rules.md` | MODIFIED — `.github/config/` row added to Directory Structure table |

## Validation Evidence

```
Tests Passed: 167, Failed: 0, Skipped: 0  (12 test files, after GH review fixes)
Quick-validate structural checks: all 0
Agents over ceiling: 0 (Code-Conductor 78, Code-Critic 55, Research-Agent 41 — all below 128)
Agent count: 14 ✓
markdownlint: 0 new errors in changed files
```

**Migration scan**: N/A (no pattern replacement migration).

## CE Gate

⏭️ CE Gate not applicable — no customer surface; this feature instruments internal agent review tooling (a measurement script + Process-Review wiring) with no user-facing interface.

## Adversarial Review Scores

| Stage | Prosecutor | Defense | Judge rulings |
|---|---|---|---|
| Code Review | 37 pts (13 sustained) | 7 pts (3 disproved, 3 rejected) | 16 rulings |
| CE Review | ⏭️ N/A | ⏭️ N/A | ⏭️ N/A |
| Post-fix Review (main) | 0 pts (0 findings) | ⏭️ N/A | pass 1: clean |
| GitHub Review (proxy) | 17 pts (5 sustained) | 6 pts (2 disproved) | 7 rulings |
| Post-fix Review (GH fixes) | 7 pts (3 sustained) | 1 pt (1 disproved) | 4 rulings |

**GitHub review intake**: 7 findings from 3 reviewers (Sourcery, Gemini, Copilot). 5 accepted, 2 rejected (GH1: CommonMark §4.2 false-positive risk; GH2: dual-purpose sentinel design verified).

## GitHub Review — Finding Summary

| ID | Source | Ruling | Notes |
|---|---|---|---|
| GH1 | Sourcery | ❌ REJECT | HeadingPattern `\s*` fix would add false positives per CommonMark §4.2 |
| GH2 | Gemini | ❌ REJECT | Dual-purpose sentinel verified — `.error` field + Count sentinel serve different consumers |
| GH3 | Gemini | ✅ ACCEPT | `compression_advisory` comment rephrased |
| GH4 | Copilot | ✅ ACCEPT (medium) | 4-backtick fence tests added (2 test cases) |
| GH5 | Copilot | ✅ ACCEPT | Stale RED-phase comment removed |
| GH6 | Copilot | ✅ ACCEPT | `pwsh \| ConvertFrom-Json` wrapped in try/catch |
| GH7 | Copilot | ✅ ACCEPT (medium) | Fence detection rewritten with depth-aware char/length tracking |

**Post-fix review (GH fixes)**: F1 ✅ tilde fence tests added; F2 ✅ indented opener test added; F3 ✅ prose clarified (express lane); F4 ❌ try/catch vs `$LASTEXITCODE` pattern distinction verified correct by defense.

## Prosecution Depth Summary

| Category | Depth | Rationale |
|---|---|---|
| architecture | full | — |
| security | full | — |
| performance | full | — |
| pattern | full | — |
| simplicity | full | — |
| script-automation | full | — |
| documentation-audit | full | — |

Re-activated categories: none

## Process Gaps Found

- Phase 2 follow-up issue created: #213 (Guidance complexity Phase 2 — retirement, structural prevention, consolidation monitoring)
- Process-Review systemic analysis: no systemic gap found (all findings were localized to the new code/guidance)

## Pipeline Metrics

Closes #211

### CODE REVIEW

File: .github/agents/Process-Review.agent.md [hunk 1]
```diff
@@ -465,11 +494,13 @@ guardrail_proposals:
     category: security
     target_file: .github/instructions/safe-operations.instructions.md
     target_section: "Section 1: File Operation Rules"
+    compression_advisory: "none — ceiling not exceeded"  # advisory text when compression_required: true; "none — ceiling not exceeded" otherwise
```
[the remark this answers is not in the document]
Comment (Grimblaz) [reply .github/agents/Process-Review.agent.md]: **Fixed** ✅ — comment rephrased in commit 2fda478. Changed "advisory text when compression_required: true; otherwise" to "advisory text if compression_required is true, otherwise" for clarity. Thank you.

**Comment by Grimblaz** [comment]
**GH1 (Sourcery — HeadingPattern regex) — Rejected** after adversarial review (high confidence).

Per CommonMark §4.2, `##Title` (no space after `#`) is not a valid ATX heading — it is parsed as a paragraph starting with `##`. Making the space optional (`^(#{2,})\s*`) would cause the script to count non-heading paragraph lines as headings, adding false positives to `section_count` and `max_nesting_depth` for any agent file that starts a line with `##` as body punctuation.

The current pattern `^(#{2,})\s` correctly matches only valid ATX headings (one required space), which is what the script needs. No change made.
